lgc/builder/BuilderImpl.cpp

```diff
@@ -117,7 +118,7 @@ Value *BuilderImplBase::CreateIntegerDotProduct(Value *vector1, Value *vector2,
       Value *product = CreateMul(elem1, elem2);
       scalar = CreateAdd(product, scalar);
```

Comment:
So your original problem, that Builder::CreateIntegerDotProduct only sometimes clamps, could be fixed by
1. Changing this to a saturating add (of the right signedness) in the case that there is an accumulator or inputs are vector of i16;
2. Documenting in Builder.h and other places that Builder::CreateIntegerDotProduct saturates the result where applicable.
Right?

However, I take your point that also extending Builder::CreateIntegerDotProduct for 32- and 64-bit inputs could reduce code duplication by the spir-v reader not having to do its own emulation.